[TIMOB-17363] Android/iOS: Added "MaskedImage" support to Android. Moved BLEND_MODE constants to "Ti.UI".

jquick-axway opened the pull request:
**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-17363

**Summary:**
- Moved `BLEND_MODE_*` constants from `Ti.UI.iOS` to `Ti.UI`.
  * Constants are still defined under `Ti.UI.iOS` for backward compatibility on iOS, but are documented as deprecated.
- Added `Ti.UI.MaskedImage` support to Android.
  * Intended as alternate solution to using rounded borders around images, which has hardware acceleration issues on Android.
- Supports 9-patch images, but only if loaded from APK's "res" directory.
- Supports base View class' `tintColor` and `opacity` properties. Applied to final result.
- Super optimized masking via GPU shaders on Android for:
  * `mask` and `tint` for all blend modes.
  * `mask` and `image` for modes `SOURCE_IN` and `DESTINATION_IN`.
- All other masking combinations not in above list resort to less optimized rendering on Android.
- Android does not support the following blend modes _(will log a warning)_:
  * `BLEND_MODE_COLOR`
  * `BLEND_MODE_COLOR_BURN`
  * `BLEND_MODE_COLOR_DODGE`
  * `BLEND_MODE_DIFFERENCE`
  * `BLEND_MODE_EXCLUSION`
  * `BLEND_MODE_HARD_LIGHT`
  * `BLEND_MODE_HUE`
  * `BLEND_MODE_LUMINOSITY`
  * `BLEND_MODE_PLUS_DARKER`
  * `BLEND_MODE_SATURATION`
  * `BLEND_MODE_SOFT_LIGHT`
- Known issues _(to be written up later)_:
  * Will not auto-rotate JPEGs based on EXIF orientation.
  * `toImage()` renders bad image on Android 7.0 and higher when masking an image.

**Test:**
Run the below test procedure on...
- Android 4.1
- Android 4.4
- Android 7.0 or higher

Test procedure...
1. Create a classic app and add the contents of "MaskedImageTest.zip" to it from [TIMOB-17363](https://jira.appcelerator.org/browse/TIMOB-17363).
2. On app startup, verify appc logo is masked within a circle like [Screenshot1](https://jira.appcelerator.org/secure/attachment/65099/Screenshot1.png)
3. Tap the "Size" button and select "Ti.UI.FILL".
4. Verify image stretches to fit container like [Screenshot2](https://jira.appcelerator.org/secure/attachment/65100/Screenshot2.png).
5. Rotate device to landscape.
6. Verify image stretches to fit landscape container like [Screenshot3](https://jira.appcelerator.org/secure/attachment/65101/Screenshot3.png).
7. Rotate device back to portrait.
8. Tap the "Mode" button and select "SOURCE_OUT".
9. Verify image is only displayed outside of circle like [Screenshot4](https://jira.appcelerator.org/secure/attachment/65102/Screenshot4.png).
10. Tap the "Mask" button and select "MaskCircleLock".
11. Tap the "Image" button and select "None".
12. Tap the "Tint" button and select "Red".
13. Verify lock is red, within a white circle, and outside of circle is red like [Screenshot5](https://jira.appcelerator.org/secure/attachment/65103/Screenshot5.png).
14. Tap the "Mode" button and select "SOURCE_IN".
15. Verify lock is white with a red circle like [Screenshot6](https://jira.appcelerator.org/secure/attachment/65104/Screenshot6.png).
16. Tap the "Size" button and select "300dp".
17. Verify circle image is no longer stretched like [Screenshot7](https://jira.appcelerator.org/secure/attachment/65105/Screenshot7.png).
18. Tap the "Image" button and select "MaskVerticalGradient".
19. Verify circle has a gradient (light red on top; full red on bottom) like [Screenshot8](https://jira.appcelerator.org/secure/attachment/65106/Screenshot8.png).

jquick-axway committed:
[TIMOB-17363] Android/iOS: Added "MaskedImage" support on Android. Moved BLEND_MODE constants from "Ti.UI.iOS" to "Ti.UI".

jquick-axway committed:
Android/iOS: Added "Ti.UI.MaskedImage" unit tests.

build commented:
<!--
  1 failure:  :disappointed_rel...
  0 warning: 
  1 messages
  
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  :disappointed_relieved: `npm test` failed. See below for details.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  ```
> titanium-mobile@7.2.0 test /Users/build/jenkins/workspace/dk_titanium_mobile_PR-10013-5LXWV5AUA4UB4MRAHHDLOLBVK56B4INHRBGUZQLHGDMKEQBWNZNA
> grunt

Running "appcJs:src" (appcJs) task

Running "eslint:src" (eslint) task

/Users/build/jenkins/workspace/dk_titanium_mobile_PR-10013-5LXWV5AUA4UB4MRAHHDLOLBVK56B4INHRBGUZQLHGDMKEQBWNZNA/build/scons-modules-integrity.js
  9:12  warning  Found non-literal argument in require  security/detect-non-literal-require

/Users/build/jenkins/workspace/dk_titanium_mobile_PR-10013-5LXWV5AUA4UB4MRAHHDLOLBVK56B4INHRBGUZQLHGDMKEQBWNZNA/android/cli/commands/_buildModule.js
  554:18  warning  Found non-literal argument to RegExp Constructor  security/detect-non-literal-regexp

/Users/build/jenkins/workspace/dk_titanium_mobile_PR-10013-5LXWV5AUA4UB4MRAHHDLOLBVK56B4INHRBGUZQLHGDMKEQBWNZNA/tests/Resources/ti.ui.maskedimage.addontest.js
   1:3   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
   2:32  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
   3:62  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
   4:57  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
   5:71  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
   6:4   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
   7:1   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
   8:1   error    Use the global form of 'use strict'                                    strict
   8:48  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
   9:1   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  10:65  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  11:10  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  12:1   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  13:31  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  14:13  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  15:25  error    Strings must use singlequote                                           quotes
  15:47  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  16:14  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  17:7   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  18:16  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  19:15  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  20:11  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  21:13  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  22:4   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  23:5   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  24:1   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  25:39  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  26:3   warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  26:46  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  27:5   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  28:1   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  29:29  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  30:40  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  31:3   warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  31:67  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  32:52  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  33:5   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  34:1   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  35:39  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  36:3   warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  36:52  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  37:3   warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  37:49  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  38:42  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  39:3   warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  39:28  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  40:3   warning  Expected an assignment or function call and instead saw an expression  no-unused-expressions
  40:36  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  41:52  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  42:56  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  43:47  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  44:61  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  45:5   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  46:1   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  47:37  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  48:22  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  49:30  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  50:36  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  51:22  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  52:16  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  53:37  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  54:22  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  55:23  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  56:7   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  57:51  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  58:66  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  59:13  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  60:6   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  61:14  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  62:5   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  63:1   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  64:38  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  65:22  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  66:30  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  67:36  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  68:22  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  69:31  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  70:37  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  71:22  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  72:23  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  73:7   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  74:51  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  75:66  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  76:13  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  77:6   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  78:14  error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  79:5   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style
  80:4   error    Expected linebreaks to be 'LF' but found 'CRLF'                        linebreak-style

/Users/build/jenkins/workspace/dk_titanium_mobile_PR-10013-5LXWV5AUA4UB4MRAHHDLOLBVK56B4INHRBGUZQLHGDMKEQBWNZNA/tests/Resources/ti.ui.tableview.addontest.js
  14:3  warning  'didFocus' is assigned a value but never used  no-unused-vars

✖ 91 problems (82 errors, 9 warnings)
  81 errors, 0 warnings potentially fixable with the `--fix` option.

Warning: Task "eslint:src" failed. Use --force to continue.

Aborted due to warnings.
npm ERR! Test failed.  See above for more details.

```
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

jquick-axway committed:
Android: Updated code formatting for [TIMOB-17363]

jquick-axway committed:
Merge branch 'master' into TIMOB-17363

jquick-axway committed:
Android/iOS: Updated formatting of MaskedImage unit test.

hansemannn commented on apidoc/Titanium/UI/UI.yml:
```diff
@@ -1203,312 +1203,250 @@ properties:
     permission: read-only
-    platforms: [iphone, ipad]
+    since: "7.2.0"
+    platforms: [android, iphone, ipad]
```
If it is not supported on Android, I would just leave out the `android` key. If we would be very exact here, we would leave the iOS only constants on the iOS namespace and put the parity ones here.

jquick-axway replied:
I still define the constant on Android, but log a warning instead. I think it's better than letting the app crash.

It is possible to add the other modes on Android, but we would have to manipulate the bitmap pixels ourselves the slow way. (yuck)

I'll remove the "android" platform keys from the unsupported constants.

hansemannn commented on iphone/Classes/UIModule.m:
```diff
@@ -164,6 +164,35 @@ - (TiUIActivityIndicatorStyleProxy *)ActivityIndicatorStyle
 MAKE_SYSTEM_PROP(PICKER_TYPE_TIME, UIDatePickerModeTime);
 MAKE_SYSTEM_PROP(PICKER_TYPE_COUNT_DOWN_TIMER, UIDatePickerModeCountDownTimer);
 
+MAKE_SYSTEM_PROP(BLEND_MODE_NORMAL, kCGBlendModeNormal);
```
Please deprecate the old ones. We have a macro for this: 
```objc
// Definition
MAKE_SYSTEM_PROP_DEPRECATED_REPLACED(name, map, api, in, newapi);

// Example
MAKE_SYSTEM_PROP_DEPRECATED_REPLACED(BLEND_MODE_NORMAL, kCGBlendModeNormal, @"UI.iOS.BLEND_MODE_NORMAL", @"7.3.0", @"UI.BLEND_MODE_NORMAL");
```

Note: The `Ti` / `Titanium` is not included into the log as it is replaced by the app name during project generation. For the log, it's manually constructed to avoid being replaced, so we leave it out in every log.

jquick-axway replied:
Will do.

hansemannn commented on tests/Resources/ti.ui.maskedimage.addontest.js:
```diff
@@ -0,0 +1,77 @@
+			image: '/SplashScreen.png',
+			mode: Ti.UI.BLEND_MODE_SOURCE_IN,
+			width: Ti.UI.FILL,
+			height: Ti.UI.FILL,
```
It should fill by default. How does this behave on iOS?

jquick-axway replied:
Android currently auto-sizes by default. I'll double check the iOS handling.

jquick-axway replied:
I just re-tested this on iOS. You're right. An iOS `MaskedImage` does a `Ti.UI.FILL` by default. I'll change Android to make it match.

Also, our iOS `MaskedImage` doesn't appear to support `Ti.UI.SIZE`. I get a 0x0 size view when attempting to auto-size it. On Android, I made this work. I auto-size based on the image if available, otherwise I fallback to auto-sizing based on the mask size. (I'll keep this behavior.)

hansemannn reviewed (changes requested):
Left some review changes, mainly about docs and iOS.

jquick-axway commented:
@hansemannn, do you know if we need to do anything on the Alloy side after moving the BLEND_MODE constants?

garymathews commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIMaskedImage.java:
```diff
@@ -0,0 +1,1399 @@
+						// Android 4.4 has a bug where if bitmap byte width does not fit a 4 byte packing alignment,
+						// then mask will appear skewed. Resizing bitmap doesn't work. So, give up if this happens.
+						if ((Build.VERSION.SDK_INT == 19) || (Build.VERSION.SDK_INT == 20)) {
+							if ((bitmap.getWidth() % 4) != 0) {
```
Could this be moved above the `bitmap.extractAlpha()` to fail before reaching `bitmap.extractAlpha()` ?

jquick-axway replied:
Ahh, good catch. I was trying to "fix" the resulting grayscaled bitmap, but later gave up.

garymathews commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIMaskedImage.java:
```diff
@@ -0,0 +1,1399 @@
+	{
+		if (this.maskedDrawable != null) {
+			if (value != null) {
+				int color = TiConvert.toColor(TiConvert.toString(value));
```
Maybe validate if `value` is a string?

garymathews commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUIMaskedImage.java:
```diff
@@ -0,0 +1,1399 @@
+	{
+		if (this.maskedDrawable != null) {
+			if (value != null) {
+				int color = TiConvert.toColor(TiConvert.toString(value));
```
Same here

jquick-axway committed:
Android/iOS: Updated "MaskedImage" changes made in [TIMOB-17363]

- Ti.UI.iOS BLEND_MODE constants will now log a warning if used.
- Modified Android's MaskedImage to be Ti.UI.FILL by default like iOS. (Was Ti.UI.SIZE)
- Now logs a warning if a non-string type was given to MaskedImage color properties.
- Modified Android to support mask, image, and tint all at the same time like iOS.
  * Used to not support both tint and image at the same time.
- Fixed dynamic changes of "mask" and "tint" properties on Android. (Would not always update onscreen.)
- Updated docs to indicate which BLEND_MODE constants are not supported by the Android platform.

jquick-axway commented:
**Updated PR:**
- `Ti.UI.iOS` BLEND_MODE constants will now log a warning if used.
- Modified Android's `MaskedImage` to be `Ti.UI.FILL` by default like iOS. (Was `Ti.UI.SIZE`.)
- Now logs a warning if a non-string type was given to `tint` and `tintColor` properties.
- Modified Android to support mask, image, and tint all at the same time like iOS.
  * Used to not support both tint and image at the same time.
- Fixed dynamic changes of `mask`, `tint`, and `mode` properties on Android. _(Would not always update onscreen.)_
- Updated docs to indicate which BLEND_MODE constants are not supported by the Android platform.

build commented:
<!--
  1 failure:  Tests have failed...
  0 warning: 
  1 messages
  1 markdown notices
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Tests have failed, see below for more information.
  </td>
    </tr>
  </tbody>
</table>



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10013/5/artifact/dist/mobilesdk-7.2.0.v20180427212322-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>

### Tests: 

| Classname | Name | Time | Error |
| --- | --- | --- | --- |
| android.Titanium.UI.Picker | TimePicker | 2.681 | AssertionError: expected 2018-04-28T04:33:00.203Z to equal 2018-04-28T04:33:59.174Z |
| android.Titanium.UI.WebView | userAgent | 10.737 | Error: timeout of 10000ms exceeded |

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

hansemannn committed:
Merge branch 'master' into TIMOB-17363

hansemannn committed:
Update UI.yml

hansemannn committed:
Update iOS.yml

hansemannn committed:
Update TiUIiOSProxy.m

build commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-10013/8/artifact/dist/mobilesdk-7.3.0.v20180517030621-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>
